src/mpl/configure.ac

```diff
@@ -684,7 +684,22 @@ AC_DEFINE_UNQUOTED([THREAD_PACKAGE_NAME],[$THREAD_PACKAGE_NAME],[set to the name
 
 # check for compiler-support for thread-local storage (MPL_TLS)
 if test "${with_thread_package}" != "argobots" ; then
-    AX_TLS([], AC_MSG_WARN([thread-local storage not supported by the compiler.]))
+    AC_MSG_CHECKING([for _Thread_local])
+    AC_LINK_IFELSE([AC_LANG_PROGRAM([extern _Thread_local int foo = 0;],[foo = 1])],
+    [AC_MSG_RESULT([yes])
+    AC_DEFINE([TLS],[_Thread_local],[Using _Thread_local for TLS])],
+    [AC_MSG_RESULT([no])
+    AC_MSG_CHECKING([for __thread])
+    AC_LINK_IFELSE([AC_LANG_PROGRAM([__thread int foo = 0;],[foo = 1])],
+    [AC_MSG_RESULT([yes])
+    AC_DEFINE([TLS],[__thread],[Using __thread for TLS])],
+    [AC_MSG_RESULT([no])
+    AC_MSG_CHECKING([for __declspec(thread)])
+    AC_LINK_IFELSE([AC_LANG_PROGRAM([__declspec(thread) int foo = 0;],[foo = 1])],
+    [AC_MSG_RESULT([yes])
+    AC_DEFINE([TLS],[__declspec(thread)],[Using __declspec(thread) for TLS])],
+    [AC_MSG_RESULT([no])
+    AC_MSG_WARN([Compiler does not support thread local storage])])])])
 fi
 
```

Comment:
Reads just like LISP :)

What I was having in mind is:
```
AC_MSG_CHECKING([for thread-local storage])
if test -z $pac_cv_tls ; then
    AC_COMPILE_IFELSE( [AC_LANG_PROGRAM([_Thread_local int foo=0;],[foo=1])],
            [pac_cv_tls=_Thread_local], []
     )
fi
if test -z $pac_cv_tls ; then 
    AC_COMPILE_IFELSE( [AC_LANG_PROGRAM([__thread int foo=0;],[foo=1])],
            [pac_cv_tls=__thread], []
     )
fi
if test -z $pac_cv_tls ; then 
    AC_COMPILE_IFELSE( [AC_LANG_PROGRAM([__declspec(thread) int foo=0;],[foo=1])],
            [pac_cv_tls="__declspec(thread)"], []
     )
fi
if -n $pac_cv_tls ; then
    AC_MSG_RESULT([no])
    # AC_MSG_WARN  or ERROR
else
    AC_MSG_RESULT([$pac_cv_tls])
    AC_DEFINE([TLS], [$pac_cv_tls], [Defined as the attribute keyword for thread-local storage.])
fi
```

Either way, this is a handful to read for a single test. Can we wrap it in `AC_DEFUN([PAC_CC_CHECK_TLS],...` in `confdb/aclocal_cc.m4`? 

PS: why do we need the `extern` for `_Thread_local` test?

Reply:
The usage in MPICH is with `extern`, so the test closely matches. I also used `AC_LINK_IFELSE` because at least one toolchain had compiler support, but broke down at the link step.

Reply:
The other suggestions here look good to me. I'll push a new version.

Reply:
Actually it looks like the link test is sufficient for what we need. `extern` doesn't affect the results, so I'm OK to remove it.